src/Cli/dotnet/commands/dotnet-workload/install/WorkloadGarbageCollector.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Deployment.DotNet.Releases;
+using Microsoft.DotNet.NativeWrapper;
+using Microsoft.NET.Sdk.WorkloadManifestReader;
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    /// <summary>
+    /// Handles garbage collection of workload sets, workload manifests, and workload packs for a feature band.
+    ///
+    /// This class does not actually do the garbage collection (ie delete / uninstall) the items.  It calculates which
+    /// items should be kept installed.  The IInstaller implementation will be responsible for actually uninstalling
+    /// the items which are not needed.  The IInstaller implementation should keep "reference counts" for workload packs
+    /// and manifests for each feature band.  When it runs garbage collection, it should remove those reference counts
+    /// for the current feature band for items not specified in ManifestsToKeep and PacksToKeep.  Then, if an item
+    /// has no reference counts left, it can actually be deleted / uninstalled.
+    /// </summary>
+    internal class WorkloadGarbageCollector
```

Comment:
Should we perhaps rename this to`FileBasedWorkloadGarbageCollector`? Looking at the code, it is primarily dealing with the install state files that are absent in other instances like MSI based ones.

Reply:
This is intended to be used for MSI-based installs also (in the future).  MSI-based installs will need the same logic of figuring out which workload sets to keep, and from that what manifests and packs should be kept.

The [install state folder](https://github.com/dotnet/sdk/blob/release/8.0.1xx/src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadInstallType.cs#L41) that it looks at will be under CommonApplicationData for MSI-based installs.  This folder can contain the following:

- Records of global.json workload pinning: https://github.com/dotnet/designs/blob/simplify-workload-versioning/accepted/2023/simplify-workload-versioning.md#side-by-side-workload-sets
- Global workload rollback / pinning information: https://github.com/dotnet/designs/blob/simplify-workload-versioning/accepted/2023/simplify-workload-versioning-implementation.md#rolling-back-and-pinning-workload-versions